sdk/nodejs/appservice/zMixins.ts

```diff
@@ -246,81 +255,257 @@ export interface HostSettings {
+     * the Azure will call into each time the FunctionApp it is is invoked.
+     */
+    callbackFactory?: CallbackFactory<C, E, R>;
+}
+
+/** @internal */
+export function serializeFunctionCallback<C extends Context<R>, E, R extends Result>(
+    args: CallbackArgs<C, E, R>): Promise<pulumi.runtime.SerializedFunction> {
 
     if (args.callback && args.callbackFactory) {
         throw new pulumi.RunError("Cannot provide both [callback] and [callbackFactory]");
     }
 
     if (!args.callback && !args.callbackFactory) {
-        throw new Error("One of [callback] or [callbackFactory] must be provided.");
+        throw new pulumi.RunError("One of [callback] or [callbackFactory] must be provided.");
     }
 
     const func = args.callback
         ? redirectConsoleOutput(args.callback)
         : () => redirectConsoleOutput(args.callbackFactory!());
 
-    const serializedFunc = pulumi.runtime.serializeFunction(
-        func, { isFactoryFunction: !!args.callbackFactory });
-
-    return pulumi.all([bindings, serializedFunc]).apply(async ([bindings, serializedFunc]) => {
-        const map: pulumi.asset.AssetMap = {};
-        map["host.json"] = new pulumi.asset.StringAsset(JSON.stringify({
-            version: "2.0",
-            tracing: { consoleLevel: "verbose" },
-            extensionBundle: {
-                id: "Microsoft.Azure.Functions.ExtensionBundle",
-                version: "[1.*, 2.0.0)"
-            },
-            ...args.hostSettings,
-        }));
+    return pulumi.runtime.serializeFunction(func, { isFactoryFunction: !!args.callbackFactory });
+}
 
-        map[`${name}/function.json`] = new pulumi.asset.StringAsset(JSON.stringify({
-            disabled: false,
-            bindings: bindings,
-        }));
+/** @internal */
+export interface FunctionArgs {
+    /**
+     * Azure Function name.
+     */
+    name: string;
 
-        map[`${name}/index.js`] = new pulumi.asset.StringAsset(`module.exports = require("./handler").handler`),
-        map[`${name}/handler.js`] = new pulumi.asset.StringAsset(serializedFunc.text);
+    /**
+     * An array of function binding definitions.
+     */
+    bindings: pulumi.Input<BindingDefinition[]>;
 
-        const pathSet = await pulumi.runtime.computeCodePaths(args.codePathOptions);
-        for (const [path, value] of pathSet.entries()) {
-            map[name + "/" + path] = value;
-        }
+    /**
+     * Serialized function callback.
+     */
+    body: Promise<pulumi.runtime.SerializedFunction>;
+
+    /**
+     * Application settings required by the function.
+     */
+    appSettings?: pulumi.Input<{ [key: string]: any }>;
 
-        return map;
-    });
+    /**
+     * HTTP route to call this function (applies to functions reachable via HTTP).
+     */
+    route?: pulumi.Input<string>;
 }
 
-function redirectConsoleOutput<C extends Context<R>, E, R extends Result>(callback: Callback<C, E, R>) {

```

Comment:
note: consider restoring the positoin of some of these helpers to make the diff easier :)